packages/grid/x-data-grid/src/components/columnHeaders/GridColumnHeaderItem.tsx

```diff
@@ -239,6 +241,7 @@ function GridColumnHeaderItem(props: GridColumnHeaderItemProps) {
       tabIndex={tabIndex}
       aria-colindex={colIndex + 1}
       aria-sort={ariaSort}
+      aria-label={column.renderHeader && headerComponent == null ? label : undefined}
```

Comment:
We could go further and always add `aria-label` when a custom rerender is provided. This can be better addressed by https://github.com/mui/mui-x/issues/4771